### DISCUSSION TITLE
"File name too long" with HTTP presigned requests

### DISCUSSION THREAD

**Discussion by cgrass** [fsspec/filesystem_spec#1490] [Q&A]
Hello,
I'm new to python and fsspec, so hopefully there is an obvious answer to my question. Thanks for the help!

### Issue
I need to fetch a file from an s3 bucket using https and a pre-signed URL. The URL produced is quite long, commonly over 1500 characters (mostly query params). If I use the code below, I get an OS exception: `File name too long`.

```
def test_this():
    kwargs = {}
    spec = fsspec.filesystem("https", **kwargs)
    frompath = {longurl}
    topath = "/var/"  ### in prod this is a real dest
    response = spec.get(frompath, topath, recursive=False, **kwargs)
    print(response)
```

I walked through the code and can see that the `rpath` is typically appended to the `lpath` for storage in `asyn.py`:
```
lpaths = other_paths(
    rpaths,
    lpath,
    exists=exists,
    flatten=not source_is_str,
)

def other_paths(
    paths: list[str],
    path2: str | list[str],
    exists: bool = False,
    flatten: bool = False,
) -> list[str]:
    if isinstance(path2, str):
        path2 = path2.rstrip("/")

        if flatten:
            path2 = ["/".join((path2, p.split("/")[-1])) for p in paths]
        else:
            cp = common_prefix(paths)
            if exists:
                cp = cp.rsplit("/", 1)[0]
            if not cp and all(not s.startswith("/") for s in paths):
                path2 = ["/".join([path2, p]) for p in paths]
            else:
                path2 = [p.replace(cp, path2, 1) for p in paths]
    else:
        assert len(paths) == len(path2)
    return path2
```

### Question
Am I misusing the lib or protocol? Is there a way to configure fsspec to scrub query params (or use some arbitrary string) for the appended destination string?

### Setup
MacOS Ventura M1 Pro
fsspec-2023.12.2
Python 3.9, 3.10 (tried both)

**Comment by martindurant** [comment]
Because your `topath` is a directory (ends with "/"), the operation will be
longurl -> "/var/" + longurl

This may not be possible to achieve, as you have found.

Instead, make your `topath` a specific *file*.

**Reply by cgrass** [reply]
Thanks for the quick response.

I tried that as well and end up with similar results. The output of `other_paths` includes the query param because only the common prefix (`cp`) is used in the `p.replace`. the subsequent line in `asyn.py` then calls `[os.makedirs(os.path.dirname(lp), exist_ok=True) for lp in lpaths]`, which uses the raw result to try to create a dir with the query param included.

**Reply by martindurant** [reply]
Ah, OK - trimming the query sounds like a reasonable thing we should be doing.
You should be able to get around this by providing your input and output paths as one-element lists.

**Reply by cgrass** [reply]
Ok, thanks. that does work if the `lpath` is a file; if `lpath` ends in a `/` and is interpreted as a dir I get the following error because the replace/append in `other_paths` is never called:
```
IsADirectoryError: [Errno 21] Is a directory: '/var/inputs/'
```

want me to create an issue for either behavior mentioned in this thread?

**Reply by martindurant** [reply]
The issue of using non-path parts in other_paths is real. I'm not sure how we would go about solving it...

**Reply by cgrass** [reply]
a lot of unit tests expect the input to be represented in the output, but is that an actual requirement? could you instead hash the full URL and use the result as a path segment in the output?
that solution should solve for collisions and result in a predictable length.

**Reply by martindurant** [reply]
This is supposed to mirror (posix) command line usage:

```bash
$ cp myfile outdir/
```
will result in "outdir/myfile"
but
```bash
$ cp myfile outfile
```
results in "outfile".

The point is, for HTTP we don't really have files, only URLs, and it's not clear whether query parameters are part of that or not (because they often do influence what content it returned). That consideration doesn't apply to (most) other storage srevices.

**Reply by cgrass** [reply]
> This is supposed to mirror (posix) command line usage

I see slightly different behavior given my use case. If `exists` is set to true, which should be true for any http `get` operation, the result would be `outfile/myfile`, as one of the unit tests demonstrates: `(["/path1"], "/path2", True, ["/path2/path1"])`.

But in either case, it seems like fsspec needs to protect against source URLs being longer than local file systems are capable of handling.

> The point is, for HTTP we don't really have files, only URLs, and it's not clear whether query parameters are part of that or not (because they often do influence what content it returned). That consideration doesn't apply to (most) other storage srevices.

The query params would still be used to fetch the remote object and then used to construct the hash that gets used for the local path. And since the full url was used to create the hash for that file, any caller should be able to reconstruct it with a simple extension method (or whatever pattern is common in python).


It sounds like you might be considering a factory or resolver pattern instead. I do think `other_paths` is doing some pretty heavy lifting, managing many different use cases.

**Reply by cgrass** [reply]
regarding the unexpected file/path behavior when `existing==true`, a colleague pointed out a behavior difference between fsspec and posix.

i would expect the following test to pass, but it fails because the output is `/path2/path3.txt/path1`:
```
(["/path1"], "/path2/path3.txt", True, ["/path2/path3.txt"])
```

I think the unexpected behavior related to the `exists` flag supports breaking up `other_paths` into file_system resolvers.

if that change feels too big, `spec.get` should probably be changed so it uses a different test for setting `exists`. it currently sets `exists=true` for URLs which creates the unexpected `/path2/path3.txt/path1` path.

**Reply by cgrass** [reply]
@martindurant i'm inclined to turn this into an issue, or maybe two: one for the long URL and one for changing behavior when calling `get` with an `lpath` that is a file and `rpath` is a URL. 

thoughts?

**Reply by martindurant** [reply]
Lodging issues is fine, it might get more notice. For the latter issue, there are quite a lot of cross-implementation batch operation tests (fsspec/tests/abstract), and perhaps best would be to find the specific case that might be missed in those, defining how the operation ought to work.

**Reply by cgrass** [reply]
ok, i created https://github.com/fsspec/filesystem_spec/issues/1504 and https://github.com/fsspec/filesystem_spec/issues/1505

**Reply by martindurant** [reply]
Feel free to close this discussion. I can't promise anyone else will get involved... If you have a possible fix, of course I would like to see it.

**Reply by cgrass** [reply]
Ok, i'll close this discussion and leave a comment in both bug reports. I'm not a python dev, so it's hard for me to recommend a specific fix, but i have a couple ideas.
